### ISSUE TITLE
KSQL creates streams with wrong Protobuf schema if the Protobuf definition has multiple messages 

### ISSUE THREAD

**Issue by lang1991** [confluentinc/ksql#5265]
**Describe the bug**
When we create a Kafka stream in ksql-cli from a Protobuf definition in the schema registry, the schema definition of the first message in the ".proto" is used, instead of the one that actually matches the content of the stream. 

Given the following Protobuf definition:
```
syntax = "proto2";

package me.kafkaplayground.service;

option java_package = "com.me.protos.kafkaplayground.service";
option java_generic_services = true;

message LogLine {
  optional string ip = 1;
  optional uint64 timestamp = 2;
  optional string url = 3;
  optional string referer = 4;
  optional string useragent = 5;
  optional User user = 6;
}

message User {
  optional string first_name = 1;
  optional string last_name = 2;
  optional int32 age = 3;
}

message DummyC {
  enum StateC {
    STATE_C_DO_NOT_USE = 0;
    ACTIVE = 1;
    INACTIVE = 2;
  }

  optional string c_token = 1;

  optional string u_token = 2;

  optional StateC state = 3;

  repeated Event events = 4;

  oneof p_data {
    BcData bb = 5;
  }

  message Event {
    optional int64 id = 1;

    oneof product_data {
      BceData b = 5;
    }

    message BceData {
      optional CState c_state = 1;
    }
  }

  message BcData {
    oneof ch_id {
      string u_token = 1;
    }

    optional string bc_token = 2;
  }
}

enum CState {
  ACTIVE = 1;
  DISABLED = 2;
  HAHA = 3;
  HEHE = 4;
  XIXI = 5;
  HUHU = 6;
  XDXD = 7;
  QQ = 8;
  EE = 9;
  A = 10;
  B = 11;
  C = 12;
  D = 13;
  E = 14;
  F = 15;
}
```
Assuming we are sending instances of DummyC through a Kafka topic "dummyC", and we create a Kafka stream in ksql-cli with the following command:
`create stream dummyC with (KAFKA_TOPIC='dummyC', VALUE_FORMAT='PROTOBUF');`
The schema of stream "dummyC" will be "Logline" in this case, because it is the first "message" defined in the ".proto" file. If we swap the order to make "DummyC" the first message in this ".proto", the stream will be created with the correct schema.

**To Reproduce**
Steps to reproduce the behavior, include:
1. The version of KSQL. **5.5.0-post**
2. Sample source data. Please refer to the Protobuf definition above
3. Any SQL statements you ran. Please refer to the `create stream` statement above.
4. Message publisher code:
```
public static void main(String[] args) throws Exception {
    Properties props = new Properties();
    props.put("bootstrap.servers", "localhost:9092");
    props.put("acks", "all");
    props.put("retries", 0);
    props.put("key.serializer", "org.apache.kafka.common.serialization.StringSerializer");
    props.put("value.serializer", "io.confluent.kafka.serializers.protobuf.KafkaProtobufSerializer");
    props.put("schema.registry.url", "http://127.0.0.1:8081");

    String topic = "dummyC";
    int messageCnt = 100;

    try (Producer<String, Service.DummyC> producer = new KafkaProducer<>(props)) {
      for(int i = 0; i < messageCnt; ++i) {
        Service.DummyC msg = newDummyC();
        ProducerRecord<String, Service.DummyC>
            record = new ProducerRecord<>(topic, UUID.randomUUID().toString(), msg);
        producer.send(record).get();
        System.out.println("Msg sent");
      }
    }
  }

  public static Service.DummyC newDummyC() {
    Service.DummyC dummyC = Service.DummyC.newBuilder()
        .setCToken("abc")
        .setUToken("123")
        .setState(Service.DummyC.StateC.ACTIVE)
        .setBb(Service.DummyC.BcData.newBuilder()
            .setBcToken("bank_token")
            .setUToken("abcd")
            .build())
        .addEvents(Service.DummyC.Event.newBuilder()
            .setId(123456789L)
            .setB(Service.DummyC.Event.BceData.newBuilder()
                .setCState(Service.CState.QQ)
                .build())
            .build())
        .build();
    return dummyC;
  }
```

**Expected behavior**
The stream should be created with the schema the message is defined with, not the first message in the ".proto" file.

**Actual behaviour**
A clear and concise description of what actually happens, including:
1. CLI output
```
ksql> create stream dummyC with (KAFKA_TOPIC='dummyC', VALUE_FORMAT='PROTOBUF');

 Message
----------------
 Stream created
----------------
ksql> describe dummyC;

Name                 : DUMMYC
 Field     | Type
----------------------------------------------------------------------------------------
 ROWTIME   | BIGINT           (system)
 ROWKEY    | VARCHAR(STRING)  (system)
 IP        | VARCHAR(STRING)
 TIMESTAMP | BIGINT
 URL       | VARCHAR(STRING)
 REFERER   | VARCHAR(STRING)
 USERAGENT | VARCHAR(STRING)
 USER      | STRUCT<FIRST_NAME VARCHAR(STRING), LAST_NAME VARCHAR(STRING), AGE INTEGER>
----------------------------------------------------------------------------------------
```
2. Error messages **None**
3. KSQL logs **No error message in the ksqldb-server container**

**Additional context**

**Comment by apurvam** [comment]
@agavra can you help triage?

**Comment by rayokota** [comment]
This is same as https://github.com/confluentinc/schema-registry/issues/1467 and has been fixed

**Comment by agavra** [comment]
I talked more with @rayokota who shed a lot of light on my lack of understanding of how protobuf deployments usually work. For those following this ticket who may not know, having multiple schemas in a single protobuf file but only actually using one of those subschemas per topic is a common development model. You would publish the same schema to multiple subjects, and only use one subschema per topic.

Given the schema that you've registered (with multiple subschemas) a topic could hypothetically have messages from any one of those types - something (multiple types in one topic) that ksqlDB doesn't support today.

That means you could be asking for one of three things:
 1. you want to be able to specify which schema within a file of many schemas ksqlDB should use. this is something that we can add in the `WITH` clause if you know your file has multiple schemas
2. you want ksqlDB to automagically determine which subschema to use. I don't think this is possible because ksqlDB is schema-driven, meaning that even if the topic has no data we need to associate a schema with it.
3. you actually want ksqlDB to support multiple message types.

The first is pretty straightforward, the second and third not so much. We'll look into this and decide how we want to tackle it going forward.

**Comment by guozhangwang** [comment]
If the subschema name is ALWAYS to be the same as the topic name (as in this example), then maybe ksqlDB can "sort of" automagically determine which subschema to use since the serde function takes in the topic name as parameters.

**Comment by apurvam** [comment]
Removing the needs triage since impact isn't clear. It's in the backlog for now.

**Comment by vcrfxia** [comment]
Linking https://github.com/confluentinc/ksql/issues/3634, which tracks option 1 laid out above by @agavra , in order to record another reason to prioritize https://github.com/confluentinc/ksql/issues/3634.

**Comment by cdadia** [comment]
+1 if this help change the priority on this feature request. This is pretty painful for folks who are using protobuf as a schema format. option 1 would be more sufficient IMHO.

**Comment by rayokota** [comment]
Materialize supports specifying the top level message name via `PROTOBUF MESSAGE <message_name>`.  See https://materialize.com/docs/sql/create-source/protobuf-kafka/#top-level-message

**Comment by rayokota** [comment]
Similar to Materialize, the syntax should be something like `WITH PROTOBUF MESSAGE <message_name>` where `<message_name>` is a fully-qualified name such as `<package_name>.<outer_message_name>.<inner_message_name>`.

In terms of implementation, at https://github.com/confluentinc/ksql/blob/master/ksqldb-serde/src/main/java/io/confluent/ksql/serde/protobuf/ProtobufSchemaTranslator.java#L41 , the call should look like 
```
ProtobufSchema protobufSchema = ((ProtobufSchema) schema).copy(messageName);
return protobufData.toConnectSchema(protobufSchema);
```

**Comment by spena** [comment]
From the 3 options, I think we should discard #2 which let ksqlDB autodetect the message name based on the topic name. If the current design is that KSQL uses the first message not matter the name, then this would cause incompatibility if we enforce that ksqlDB matches the topic name with the protobuf message name. We instead need to pass the <message_name> in the WITH clause.

I was thinking on new WITH properties for protobuf, but I then found these two ones available in the code:
```
KEY_SCHEMA_FULL_NAME
VALUE_SCHEMA_FULL_NAME
VALUE_AVRO_SCHEMA_FULL_NAME
```
I am not sure how they're use for. They were added in KLIP-56 and mentioned this:
```
Support `KEY_SCHEMA_FULL_NAME` and `VALUE_SCHEMA_FULL_NAME` properties to control full schema name for _AVRO_ and _PROTOBUF_ format
```

I think I can re-use those configs names to set the protobuf message name. They do not work right now for Protobuf. I did a quick prototype to re-use them and things worked. @lihaosky Can I use those configs to specify the Protobuf message name or are they reserved for something else?
